### ISSUE TITLE
Audit imports corner cases

### ISSUE THREAD

**Issue by Gankra** [mozilla/cargo-vet#4]
Some corners with foreign imports that I noted during skeleton implementation:

* [ ] What should we do if some of the auditfiles fail to download/parse? 
  * Currently I emit a warning and proceed without them
  * Should we then try to use the old copy in the imports.lock? (merge)
  * Should we discard all downloads and only use imports.lock? (fallback)
  
* [ ] Should we try to hold on to "old" copies of foreign auditfiles?
  * I currently just overwrite the imports.lock completely
    * This is probably Objectively Wrong with my above behaviour for failed downloads/parses? 
  * Matters for situations where the foreign auditfile removes an entry, do we respect that, or remember that it existed?
  
* [ ] When should newly-downloaded imports be stored in imports.lock?
  * I currently do this after a successful vet, so no state is modified if the vet fails.
  
* [ ] What if an auditfile is in the lock but it's been removed from the config (with or without --locked)? 
  * I don't currently check for this at all, we either download everything in the config or just use the contents of imports.lock
  * In writing this I just realized there is a nasty corner case in my impl where you remove all imports and we keep using the old lock because there's nothing to download...

**Comment by Gankra** [comment]
> Once you start recording audits, you should generally avoid making changes to the default criteria, since such changes will automatically apply to the bulk of your previously-recorded audits and likely misrepresent what actually happened.

**Comment by Gankra** [comment]
>  The descriptions for any mapped criteria are stored as well, and any changes to previously-recorded descriptions will cause cargo vet to fail and require cargo vet accept-criteria-change to be run. Any unmapped criteria (and audits under those criteria) are discarded.

**Comment by bholley** [comment]
> Some corners with foreign imports that I noted during skeleton implementation:
> 
>     * [ ]  What should we do if some of the auditfiles fail to download/parse?
>       
>       * Currently I emit a warning and proceed without them
>       * Should we then try to use the old copy in the imports.lock? (merge)
>       * Should we discard all downloads and only use imports.lock? (fallback)

For now I think we should hard-fail. Any integration into a production build system will pass --locked to avoid having offline builds and CI hit the network in a non-determinstic way. The semantics of bare `cargo vet` involve fetching updates, so if those updates aren't available I think it's most appropriate for the tool to fail (similar to crates.io not being available).

>     * [ ]  Should we try to hold on to "old" copies of foreign auditfiles?
>       
>       * I currently just overwrite the imports.lock completely
>         
>         * This is probably Objectively Wrong with my above behaviour for failed downloads/parses?

I think just replacing imports.lock with a new copy is fine. If we early-abort on failed fetches, we won't be replacing a working imports.lock with a partial one, and it should be checked into version control anyway.

>       * Matters for situations where the foreign auditfile removes an entry, do we respect that, or remember that it existed?

We have to respect the removal, because that's the only way audits can be revoked.

>     * [ ]  When should newly-downloaded imports be stored in imports.lock?
>       
>       * I currently do this after a successful vet, so no state is modified if the vet fails.

I'm unsure as to whether it's better to do what you describe or to do it after successful fetch (but before starting the actual checking), but I think what you describe is fine for now.

>     * [ ]  What if an auditfile is in the lock but it's been removed from the config (with or without --locked)?
>       
>       * I don't currently check for this at all, we either download everything in the config or just use the contents of imports.lock

Yes, I think this is right.

>       * In writing this I just realized there is a nasty corner case in my impl where you remove all imports and we keep using the old lock because there's nothing to download...

 Conceptually, there are two separate tasks: (1) fetch and cache the new state of everyone's audits (skipped with --locked), and (2) do the checking. Ideally the code would be structured to keep these things mostly separate and allow us to skip either as requested.

**Comment by Gankra** [comment]
This all seems reasonable, but I do think there probably should be some kind of warning/error for this case:

> What if an auditfile is in the lock but it's been removed from the config (with ~~or without~~ --locked)?

Like you don't want someone revoking an import in the config and then forgetting to regenerate the lock, causing the CI to happily keep using that "bad" import?

**Comment by bholley** [comment]
I agree it's reasonable for `cargo vet --locked` to fail if the imports in `config.toml` don't match the ones in `imports.lock`.

That said, I don't expect re-configuring imports will be a common task, so I don't think it's a huge priority.

**Comment by Gankra** [comment]
closing in favour of #240
